mdn/src/main/java/org/apache/james/mdn/MDN.java

```diff
@@ -74,10 +83,83 @@ public MDN build() {
         }
     }
 
+    public static class MDNParseException extends Exception {
+        public MDNParseException(String message) {
+            super(message);
+        }
+
+        public MDNParseException(String message, Throwable cause) {
+            super(message, cause);
+        }
+    }
+
+    public static class MDNParseContentTypeException extends MDNParseException {
+        public MDNParseContentTypeException(String message) {
+            super(message);
+        }
+    }
+
+    public static class MDNParseBodyPartInvalidException extends MDNParseException {
+
+        public MDNParseBodyPartInvalidException(String message) {
+            super(message);
+        }
+    }
+
+
     public static Builder builder() {
         return new Builder();
     }
 
+    public static MDN parse(Message message) throws MDNParseException {
+        if (!message.isMultipart()) {
+            throw new MDNParseContentTypeException("MDN Message must be multipart");
+        }
+        List<Entity> bodyParts = ((Multipart) message.getBody()).getBodyParts();
+        if (bodyParts.size() < 2) {
+            throw new MDNParseBodyPartInvalidException("MDN Message must contain at least two parts");
+        }
+        try {
+            var humanReadableTextEntity = bodyParts.get(0);
+            return extractHumanReadableText(humanReadableTextEntity)
+                    .flatMap(humanReadableText1 -> extractMDNReport(bodyParts.get(1))
+                            .map(report1 -> MDN.builder()
```

Comment:
why do you need to put a 1 as a variable suffix here? Just call it `report` and also `humanReadableText` for the one above